rfd/0001-testing-guidelines.md

```diff
@@ -0,0 +1,197 @@
+---
+authors: Andrew Lytvynov (andrew@gravitational.com)
+state: draft
+---
+
+# RFD 1 - Testing guidelines
+
+## What
+
+Guidelines for writing Go tests for teleport.
+
+## Why
+
+At the time of writing this, teleport tests use
+[gocheck](https://labix.org/gocheck). This framework has several downsides:
+- all tests are serial - slow execution in packages with many small tests
+- easy to mess up setup (forgetting to call `check.T`) or run tests multiple
+  times (calling `check.T` multiple times)
+- limited deep-diffing of values
+- largely unmaintained
+
+At the same time, there are some features we'd like to retain:
+- organization of tests into testsuites
+- shared setup/teardown code between tests
+- asserts for concise validation
+
+Go's builtin `testing` package covers most of our needs, and gets new features
+with every new Go release. We just need to complement it with:
+- asserts
+- flexible value diffing (got vs want)
+
+## Details
+
+Use the `testing` package as a base and complement it with several libraries.
+
+### Test organization
+
+Use a hierarchy of:
+- `_test.go` files
+- `func Test*` functions
+- [`t.Run`](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks)
+  subtests (aka table-driven tests)
+
+Shared setup/teardown:
+- [`func TestMain`](https://golang.org/pkg/testing/#hdr-Main) for package level
+- a single `func Test*` with subtests for smaller test groups
+- `func setupX` called from multiple `func Test*` for larger test groups
+
+For performance testing use
+[benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks).
+
+### Parallelization
+
+Call [`t.Parallel()`](https://golang.org/pkg/testing/#T.Parallel) from all
+tests and subtests that don't need to run serially.
+
+### Asserts
+
+Use
+[testify/assert](https://pkg.go.dev/github.com/stretchr/testify/assert?tab=doc)
+or a plain `if` condition, whichever is easiest.
+
+Skim through the `testify/assert` docs, it has many convenient helpers.
+
+### Diffing
+
+For non-trivial comparisons (such as deeply-nested structs or protobufs), use
+[go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp?tab=doc) with
```

Comment:
@awly it seems that go-cmp has many similar features, do we need it?

Reply:
My preference is using `go-cmp` and plain `ifs` + `t.Error/Fatal`. `testify/assert` is definitely not a must.
I know that at least @russjones prefers the assert syntax for shorter code (avoiding `ifs`).

@webvictim @fspmarshall @kimlisa @alex-kovoy @quinqu what are your thoughts on asserts vs ifs?

Reply:
I have to admit that I really like the readability granted by the `testify/assert` package.  Especially with boiler-plate stuff like `assert.Contains` or `assert.InDelta`.  I also don't really see a downside of using `testify/assert`, since its well maintained and isn't a framework (i.e. doesn't dictate broader test structure).

Reply:
I have no big preference either way - I'll go with whatever others decide.

Reply:
Sounds like we're slightly leaning towards `testify/assert` as a group without any pushback.
`go-cmp` is a transient dependency anyway, so we can keep it as an option.
Added a suggestion in this section to reach for `testifya/assert` first and fall back to `go-cmp` for non-trivial cases.